File: frontend/src/lib/calc/loan-math.ts
import type { DpdBucket } from "@/lib/domain/collections";
import type { LoanCostBreakdown } from "@/lib/domain/loan";

/**
 * Loan math for the salary-linked single-repayment product.
 *
 * Worked example for a principal of ₹10,000 over 30 days:
 *  - processingFee  = 10,000 * 0.10            = ₹1,000
 *  - gstOnFee       = 1,000 * 0.18             = ₹180
 *  - netDisbursed   = 10,000 - 1,000 - 180     = ₹8,820
 *  - interest (30d) = 10,000 * 0.01 * 30       = ₹3,000
 *  - totalRepayable = 10,000 + 3,000           = ₹13,000
 *    (processing fee + GST are taken up-front from disbursal, not added to
 *     the repayment; repayment = principal + accrued interest)
 */

/** Up-front processing fee rate on principal. */
export const PROCESSING_FEE_RATE = 0.1;
/** GST rate applied to the processing fee. */
export const GST_RATE = 0.18;
/** Interest accrual per day on principal. */
export const DAILY_INTEREST_RATE = 0.01;
/** Late penalty per day after the due date. */
export const LATE_PENALTY_RATE = 0.02;
/** Maximum days the late penalty accrues before collections. */
export const LATE_PENALTY_CAP_DAYS = 30;
/** Flat instant-loan ceiling: ₹10,00,000 (salary no longer caps the amount). */
export const MAX_INSTANT_LOAN_AMOUNT = 1_000_000;

/** Floor on a sanctioned advance (working value — finalised in credit policy). */
export const MIN_LOAN_AMOUNT = 1000;

/** Round to whole rupees (amounts are presented and settled in INR). */
function inr(value: number): number {
  return Math.round(value);
}

/** Up-front 10% processing fee on the principal. */
export function processingFee(amount: number): number {
  return inr(amount * PROCESSING_FEE_RATE);
}

/** 18% GST charged on the processing fee. */
export function gstOnFee(fee: number): number {
  return inr(fee * GST_RATE);
}

/** Amount credited to the borrower = principal - fee - GST. */
export function netDisbursed(amount: number): number {
  const fee = processingFee(amount);
  return inr(amount - fee - gstOnFee(fee));
}

/** Interest accrued over `days` at 1%/day on principal. */
export function dailyInterest(amount: number, days: number): number {
  return inr(amount * DAILY_INTEREST_RATE * Math.max(0, days));
}

/** Total amount due on the repayment date = principal + interest. */
export function totalRepayable(amount: number, days: number): number {
  return inr(amount + dailyInterest(amount, days));
}

/** Late penalty accrued for `daysLate`, capped at 30 days. */
export function latePenalty(amount: number, daysLate: number): number {
  const cappedDays = Math.min(Math.max(0, daysLate), LATE_PENALTY_CAP_DAYS);
  return inr(amount * LATE_PENALTY_RATE * cappedDays);
}

/** Eligible loan limit — a flat ₹10,00,000 instant cap (salary drives the due date, not the amount). */
export function eligibleLimit(_monthlySalary: number): number {
  return MAX_INSTANT_LOAN_AMOUNT;
}

/** Map days-past-due to a {@link DpdBucket}. */
export function dpdBucket(daysPastDue: number): DpdBucket {
  if (daysPastDue <= 0) return "UPCOMING";
  if (daysPastDue <= 7) return "T0_T7";
  if (daysPastDue <= 30) return "T8_T30";
  if (daysPastDue <= 60) return "T30_T60";
  if (daysPastDue <= 90) return "T60_T90";
  return "T90_PLUS";
}

/** Whole days between two dates (b - a), ignoring time-of-day. */
export function daysBetween(a: Date, b: Date): number {
  const ms = 24 * 60 * 60 * 1000;
  const da = Date.UTC(a.getFullYear(), a.getMonth(), a.getDate());
  const db = Date.UTC(b.getFullYear(), b.getMonth(), b.getDate());
  return Math.round((db - da) / ms);
}

/** Salary-due-date window working values (spec: exact rule TBD). */
export const SALARY_DUE_MIN_CYCLE_DAYS = 15;
export const SALARY_DUE_MAX_WINDOW_DAYS = 40;

/** Clamp a day-of-month to a valid date within the given month. */
function salaryDateInMonth(year: number, monthIndex: number, salaryDay: number): Date {
  const lastDay = new Date(year, monthIndex + 1, 0).getDate();
  return new Date(year, monthIndex, Math.min(salaryDay, lastDay));
}

/**
 * Derive the single-repayment due date from the borrower's salary credit.
 * Repayment lands on the salary day — the next salary credit that gives a
 * reasonable cycle (>= ~15 days) while staying within the ~40-day window.
 *
 * Examples (salary on 30th): disbursed 3 Jun → due 30 Jun (~27d);
 * disbursed 25 Jun → 30 Jun is only ~5d out, so roll to 30 Jul (~35d).
 */
export function dueDateFromSalary(params: {
  disbursedOn: Date;
  /** Day-of-month the salary is typically credited. */
  salaryDay: number;
  /** Recent salary credit dates observed (bank statement). */
  recentSalaryDates?: Date[];
}): Date {
  const { disbursedOn, salaryDay } = params;
  let candidate = salaryDateInMonth(
    disbursedOn.getFullYear(),
    disbursedOn.getMonth(),
    salaryDay,
  );

  // Roll forward until the candidate is strictly after disbursement and gives
  // at least a minimum cycle, without exceeding the max window.
  for (let i = 0; i < 3; i += 1) {
    const gap = daysBetween(disbursedOn, candidate);
    if (gap >= SALARY_DUE_MIN_CYCLE_DAYS) break;
    const next = new Date(candidate);
    next.setMonth(next.getMonth() + 1);
    const rolled = salaryDateInMonth(next.getFullYear(), next.getMonth(), salaryDay);
    if (daysBetween(disbursedOn, rolled) > SALARY_DUE_MAX_WINDOW_DAYS) break;
    candidate = rolled;
  }
  return candidate;
}

/**
 * Build a full itemized {@link LoanCostBreakdown} for a principal and tenure.
 */
export function buildCostBreakdown(amount: number, tenureDays: number): LoanCostBreakdown {
  const fee = processingFee(amount);
  const gst = gstOnFee(fee);
  const interest = dailyInterest(amount, tenureDays);
  return {
    principal: amount,
    processingFee: fee,
    gstOnFee: gst,
    netDisbursed: inr(amount - fee - gst),
    interest,
    tenureDays,
    totalRepayable: inr(amount + interest),
  };
}

File: backend/navix-loan/src/main/java/com/navix/loan/service/LoanMath.java
package com.navix.loan.service;

import java.math.BigDecimal;
import java.math.RoundingMode;
import java.time.LocalDate;
import org.springframework.stereotype.Service;

/**
 * Pure money-math for the salary-linked single-repayment loan product.
 *
 * <p><b>Money is integer paise everywhere</b> (1 rupee = 100 paise). Every amount in and
 * out of this class is a {@code long} number of paise; there is no floating-point money.
 * Rounding to a whole paisa is always {@link RoundingMode#HALF_UP}. This class is the
 * backend source of truth for loan economics and mirrors
 * {@code frontend/src/lib/calc/loan-math.ts}.
 *
 * <p>Pricing is flat across risk categories. Worked example for a principal of ₹10,000
 * (= 1_000_000 paise):
 * <pre>
 *   processingFee  = 1_000_000 * 0.10          =   100_000   (₹1,000)
 *   gst            =   100_000 * 0.18           =    18_000   (₹180)
 *   netDisbursed   = 1_000_000 - 100_000 - 18_000 = 882_000   (₹8,820)
 *   interest (30d) = 1_000_000 * 0.01 * 30       =   300_000   (₹3,000)
 *   totalRepayable = 1_000_000 + 300_000         = 1_300_000   (₹13,000)
 * </pre>
 * Processing fee + GST are taken up front (they reduce {@code netDisbursed}); they are
 * <i>not</i> part of {@code totalRepayable}, which is principal + accrued interest.
 */
@Service
public class LoanMath {

    /** Up-front processing fee rate: 10% of principal. */
    public static final BigDecimal PROCESSING_FEE_RATE = new BigDecimal("0.10");

    /** GST rate applied on the processing fee: 18%. */
    public static final BigDecimal GST_RATE = new BigDecimal("0.18");

    /** Interest accrued per day: 1% of principal. */
    public static final BigDecimal DAILY_INTEREST_RATE = new BigDecimal("0.01");

    /** Late penalty per day past due: 2% of principal. */
    public static final BigDecimal LATE_PENALTY_RATE = new BigDecimal("0.02");

    /** Late penalty accrues for at most this many days, then collections. */
    public static final int LATE_PENALTY_CAP_DAYS = 30;

    /** Flat instant-loan ceiling: ₹10,00,000 = 100,000,000 paise (salary no longer caps the amount). */
    public static final long MAX_INSTANT_LOAN_PAISE = 100_000_000L;

    /** Eligible limit is rounded down to the nearest ₹100 (= 10,000 paise). */
    public static final long LIMIT_ROUNDING_PAISE = 10_000L;

    /** Maximum loan term: the due date must fall within this many days of disbursement. */
    public static final int MAX_TERM_DAYS = 40;

    /** Fixed loan term: 30-day single-repayment bullet loan, due = disbursed + 30d (dfd.md D10). */
    public static final int TERM_DAYS = 30;

    /** Grace after the due date before any penalty accrues (paying the day after salary is free). */
    public static final int SALARY_GRACE_DAYS = 1;

    /** Floor on a sanctioned advance: ₹1,000 = 100,000 paise. */
    public static final long MIN_LOAN_PAISE = 100_000L;

    /** Up-front processing fee = round(principal × 10%). */
    public long processingFeePaise(long principalPaise) {
        return roundPaise(BigDecimal.valueOf(principalPaise).multiply(PROCESSING_FEE_RATE));
    }

    /** GST on the processing fee = round(fee × 18%). */
    public long gstPaise(long principalPaise) {
        return roundPaise(BigDecimal.valueOf(processingFeePaise(principalPaise)).multiply(GST_RATE));
    }

    /** Net amount credited to the borrower = principal − fee − GST. */
    public long netDisbursedPaise(long principalPaise) {
        return principalPaise - processingFeePaise(principalPaise) - gstPaise(principalPaise);
    }

    /** Interest accrued over {@code days} = round(principal × 1%/day × days). Negative days clamp to 0. */
    public long interestPaise(long principalPaise, int days) {
        int d = Math.max(0, days);
        return roundPaise(BigDecimal.valueOf(principalPaise)
                .multiply(DAILY_INTEREST_RATE)
                .multiply(BigDecimal.valueOf(d)));
    }

    /** Total repayable after {@code days} held = principal + accrued interest (fee/GST excluded). */
    public long totalRepayablePaise(long principalPaise, int days) {
        return principalPaise + interestPaise(principalPaise, days);
    }

    /**
     * Late penalty for {@code daysLate} = round(principal × 2%/day × min(daysLate, 30)).
     * Non-positive {@code daysLate} yields 0.
     */
    public long latePenaltyPaise(long principalPaise, int daysLate) {
        int cappedDays = Math.min(Math.max(0, daysLate), LATE_PENALTY_CAP_DAYS);
        return roundPaise(BigDecimal.valueOf(principalPaise)
                .multiply(LATE_PENALTY_RATE)
                .multiply(BigDecimal.valueOf(cappedDays)));
    }

    /**
     * Eligible loan limit — a flat ₹10,00,000 for every eligible borrower (instant-loan model).
     * Salary no longer caps the amount; it still drives the due date.
     *
     * @param monthlySalaryPaise gross monthly salary in paise (unused; kept for signature stability)
     * @return the flat sanctionable limit in paise
     */
    public long eligibleLimitPaise(long monthlySalaryPaise) {
        return MAX_INSTANT_LOAN_PAISE;
    }

    /**
     * Canonical compute-on-read outstanding balance as of some day.
     *
     * <p>{@code outstanding = principal + interest(daysHeld) + penalty(daysLate) − payments}.
     * Prepayment falls out naturally: pass the actual days held so interest is charged only to
     * the day paid. {@code daysLate} is 0 until the due date (the 1-day grace is applied by the
     * caller, e.g. paying the day after salary passes {@code daysLate = 0}).
     *
     * @param principalPaise principal in paise
     * @param daysHeld       whole days from disbursement to the as-of date (interest accrues)
     * @param daysLate       whole days past the due date after grace (penalty accrues, capped)
     * @param paymentsPaise  sum of verified payments already made, in paise
     * @return remaining amount owed in paise (never negative)
     */
    public long outstandingPaise(long principalPaise, int daysHeld, int daysLate, long paymentsPaise) {
        long gross = totalRepayablePaise(principalPaise, daysHeld)
                + latePenaltyPaise(principalPaise, daysLate);
        return Math.max(0L, gross - paymentsPaise);
    }

    /**
     * Whole days a loan is past due as of {@code asOf} (0 before/on the due date).
     * Bucketing into DPD buckets lives in the collections module.
     */
    public int daysPastDue(LocalDate dueDate, LocalDate asOf) {
        long days = dueDate.until(asOf, java.time.temporal.ChronoUnit.DAYS);
        return (int) Math.max(0L, days);
    }

    /**
     * Single-repayment due date: the <b>latest</b> salary-credit date that is strictly after
     * disbursement and falls within {@link #MAX_TERM_DAYS} days of it — i.e. the last salary the
     * borrower receives inside the 40-day window (maximising tenure up to 40 days).
     *
     * <p>The salary day is clamped to each month's length (e.g. day 31 → 30 Apr, 28/29 Feb).
     * Within any 40-day window there is always at least one monthly salary date.
     *
     * @param disbursedOn     date of disbursement
     * @param salaryCreditDay day of month salary is credited (1–31)
     * @return the computed due date
     */
    public LocalDate dueDateFromSalary(LocalDate disbursedOn, int salaryCreditDay) {
        LocalDate windowEnd = disbursedOn.plusDays(MAX_TERM_DAYS);
        LocalDate best = null;
        // Walk salary dates from the disbursement month through the window-end month.
        for (LocalDate month = disbursedOn.withDayOfMonth(1);
                !month.isAfter(windowEnd.withDayOfMonth(1));
                month = month.plusMonths(1)) {
            LocalDate salaryDate = salaryDateInMonth(month, salaryCreditDay);
            if (salaryDate.isAfter(disbursedOn) && !salaryDate.isAfter(windowEnd)) {
                if (best == null || salaryDate.isAfter(best)) {
                    best = salaryDate;
                }
            }
        }
        // Fallback (should not happen for a monthly salary, but keep it total): the first salary
        // strictly after disbursement, even if it exceeds the window.
        if (best == null) {
            LocalDate salaryThisMonth = salaryDateInMonth(disbursedOn, salaryCreditDay);
            best = salaryThisMonth.isAfter(disbursedOn)
                    ? salaryThisMonth
                    : salaryDateInMonth(disbursedOn.plusMonths(1), salaryCreditDay);
        }
        return best;
    }

    /** Clamp a salary day-of-month to a valid date within the month of {@code anyDayInMonth}. */
    private static LocalDate salaryDateInMonth(LocalDate anyDayInMonth, int salaryCreditDay) {
        int lastDay = anyDayInMonth.lengthOfMonth();
        int day = Math.min(Math.max(1, salaryCreditDay), lastDay);
        return anyDayInMonth.withDayOfMonth(day);
    }

    /** Round fractional paise to a whole paisa, HALF_UP. */
    private static long roundPaise(BigDecimal paise) {
        return paise.setScale(0, RoundingMode.HALF_UP).longValueExact();
    }
}
